gateway_lambda/index.ts

```diff
@@ -0,0 +1,146 @@
+import { sendMessage } from './sqs';
+import { SqsEvent } from './types';
+
+Sentry.AWSLambda.init({
+  dsn: config.sentry.dsn,
+  release: config.sentry.release,
+  environment: config.environment,
+});
+
+/**
+ * Format the response data
+ * Source: https://docs.aws.amazon.com/lambda/latest/dg/services-apigateway.html#apigateway-types-transforms
+ * @param statusCode
+ * @param message
+ * @param isError
+ */
+export function formatResponse(
+  statusCode,
+  message: string,
+  isError?: boolean
+): APIGatewayProxyResult {
+  return {
+    statusCode,
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify({
+      statusCode,
+      [isError ? 'error' : 'message']: message,
+    }),
+  };
+}
+
+/**
+ * Generate the data for the SQS messages
+ * @param data
+ */
+export function generateEvents(data): SqsEvent[] {
+  const userId = data.payload.sub;
+  const events = data.payload.events;
+  const allowedEvents = config.fxa.allowedEvents;
+
+  return Object.keys(events)
+    .filter((event) => Object.keys(allowedEvents).includes(event))
+    .map((event) => ({
+      user_id: userId,
+      event: allowedEvents[event],
+      timestamp: Math.round(new Date().getTime() / 1000),
+    }));
+}
+
+/**
+ * Create the message for a successful execution
+ * If messages fail, update message to point to
+ * where the logs are located.
+ * @param totalEvents
+ * @param failedEvents
+ */
+export function createSuccessResponseMessage(
+  totalEvents: number,
+  failedEvents: number
+): string {
+  const numberOfSuccessfulEvents = totalEvents - failedEvents;
+  let message = `Successfully sent ${numberOfSuccessfulEvents} out of ${totalEvents} events to SQS.`;
+  if (totalEvents > numberOfSuccessfulEvents) {
+    message +=
+      ' Review cloudwatch and sentry logs for information about failed events.';
+  }
+
+  return message;
+}
+
+/**
+ * Handle API Gateway Fxa Events
+ * Endpoint: /events
+ * @param event
+ */
+export async function eventHandler(
+  event: APIGatewayEvent
+): Promise<APIGatewayProxyResult> {
+  // Get the authorization header
+  const authHeader = event.headers.authorization;
+  if (!authHeader) {
+    return formatResponse(400, 'Missing authorization header', true);
+  }
+
+  // Split the header to get the parts
+  const authHeaderParts = authHeader.split(' ');
+
+  // Make sure it's the correct auth type the first part of the header
+  // should be "Bearer"
+  if (authHeaderParts[0] !== 'Bearer') {
+    return formatResponse(401, 'Invalid auth type', true);
+  }
+  // Decode the authorization header
+  let data;
+  try {
+    data = await validate(authHeaderParts[1]);
+  } catch (error) {
+    return formatResponse(401, error.message, true);
+  }
+
+  // Generate SQS messages
+  const sqsEvents = generateEvents(data);
+  const totalEvents = sqsEvents.length;
+
+  // If there isn't any valid events, respond early
+  if (!totalEvents) {
+    return formatResponse(200, 'No valid events');
```

Comment:
it is 404 and not a success..so the firefox side can have an alert if they are sending wrong events.

Reply:
Hmm. I think it's ok since we're building this service based on *their* webhook events. We just only want to process a subset of them.

Reply:
We are not listening to all events right now, we are only interested in the `profile change` and `user delete events`. If an event we are not consuming comes through, I don't think we should be sending a not-found response back to FxA. I could be convinced otherwise